cpp/src/join/semi_join.cu

```diff
@@ -71,65 +92,71 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
-  auto const row_bitmask = (compare_nulls == null_equality::EQUAL)
-                             ? rmm::device_buffer{}
-                             : cudf::detail::bitmask_and(right_flattened_keys, stream).first;
   // skip rows that are null here.
-  thrust::for_each_n(
-    rmm::exec_policy(stream),
-    thrust::make_counting_iterator<size_type>(0),
-    right_num_rows,
-    [hash_table, row_bitmask = static_cast<bitmask_type const*>(row_bitmask.data())] __device__(
-      size_type idx) mutable {
-      if (!row_bitmask || cudf::bit_is_set(row_bitmask, idx)) {
-        hash_table.insert(thrust::make_pair(idx, true));
-      }
-    });
+  if ((compare_nulls == null_equality::EQUAL) or (not nullable(right_keys))) {
+    // TODO: Need to enable a stream argument in cuco, see
+    // https://github.com/NVIDIA/cuCollections/pull/113
+    hash_table.insert(iter, iter + right_num_rows, hash_build, equality_build);
+    // hash_table.insert(iter, iter + right_num_rows, hash_build, equality_build, stream.value());
+  } else {
+    thrust::counting_iterator<size_type> stencil(0);
+    [[maybe_unused]] auto const [row_bitmask, null_count] =
+      cudf::detail::bitmask_and(right_flattened_keys, stream);
+    row_is_valid pred{static_cast<bitmask_type const*>(row_bitmask.data())};
+
+    // insert valid rows
+    // TODO: Need to enable a stream argument in cuco, see
+    // https://github.com/NVIDIA/cuCollections/pull/113
+    hash_table.insert_if(iter, iter + right_num_rows, stencil, pred, hash_build, equality_build);
+    // hash_table.insert_if(iter, iter + right_num_rows, stencil, pred, hash_build, equality_build,
+    // stream.value());
+  }
 
-  //
   // Now we have a hash table, we need to iterate over the rows of the left table
   // and check to see if they are contained in the hash table
-  //
+  auto left_rows_d = table_device_view::create(left_flattened_keys, stream);
+  row_hash hash_probe{*left_rows_d};
+  // Note: This equality comparator violates symmetry of equality and is
+  // therefore relying on the implementation detail of the order in which its
+  // operator is invoked. If cuco makes no promises about the order of
+  // invocation this seems a bit unsafe.
+  row_equality equality_probe{*right_rows_d, *left_rows_d, compare_nulls == null_equality::EQUAL};
 
   // For semi join we want contains to be true, for anti join we want contains to be false
   bool const join_type_boolean = (kind == join_kind::LEFT_SEMI_JOIN);
 
+  auto hash_table_view = hash_table.get_device_view();
+
   auto gather_map =
     std::make_unique<rmm::device_uvector<cudf::size_type>>(left_num_rows, stream, mr);
 
   // gather_map_end will be the end of valid data in gather_map
   auto gather_map_end = thrust::copy_if(
     rmm::exec_policy(stream),
-    thrust::make_counting_iterator<size_type>(0),
-    thrust::make_counting_iterator<size_type>(left_num_rows),
+    thrust::make_counting_iterator(0),
```

Comment:
Hm, this actually should just be a `retrieve` and a new algorithm that is the inverse of `retrieve`.

Reply:
After our discussions online I think we should punt on adding further functionality to static_map for now and just use this device API until we have a chance to move forward with rapidsai/cuCollections#110.